### ISSUE TITLE
Middle Eastern languages (Arabic) internationalization

### ISSUE THREAD

**Issue by ggarra13** [fltk/fltk#1063]
**Is your feature request related to a problem? Please describe.**

FLTK 1.4.0 supports internationalization through gettext.  However, for Arabic and other middle eastern languages, which are one of the 5 most popular languages in the world, FLTK offers no provision for it.
The issue with Arabic is that it is read from right to left (that's no issue with gettext).  But... icons and buttons layouts need to also be flipped, and that's where FLTK fails.

**I am using FLTK**
 - 1.4.0

**Describe the solution you'd like**
Extend Fl_Flex, Fl_Group and Fl_Pack with a:

1) flag to revert horizontal orientation of its items.
2) option to set 1) based on setlocale().

to revert the drawing direction of its items when they are laid horizontally.  Text is already reversed if you use libintl/gettext.

**Describe alternatives you've considered**
None.  Currently supporting Arabic and other middle Eastern languages with FLTK is broken.

**Additional context**

Here's a web link with an in-depth description of the UI layout designs for middle eastern languages: 
https://medium.com/blackboard-design/fundamentals-of-right-to-left-ui-design-for-middle-eastern-languages-afa7663f66ed

**Comment by MatthiasWM** [comment]
One thing ahead of the long explanation: you can easily align labels to the right and put icons to the right of labels for many widgets. Notable exceptions are menu items and content in editor and limited in browser lines (right-aligning works though).

Phew, this is a huge task to do correctly. FLTK was originally written without much concern even for Western scripts besides U.S. English. Bill supported some Emacs style shortcuts, I added some support for German which was later expanded to very many left-to-right scripts, including input methods for Korean and Chinese and more.

Implementing right-to-left script is a whole different number on FLTK. All text editors would need to be fundamentally rewritten, especially if we allow users to mix right-to-left and left-to-right in the same line. Three things need to come together for such fundamental changes in a free OpenSource library: demand, expertise, and consent.

Is there demand from a big developer group, or by an app with a big user group? If adding this code does not increase the user base significantly, it will be impossible to convince existing developers to implement this.

Is there expertise to write such a thing? It would be impossible to write code working with middle eastern script without knowing and understanding the languages and the peculiarities of the script. If there is a fresh developer to add this kind of code, it may be accepted...

There must also be consent among the current developers that adding and restructuring so much code in the code base is acceptable and will not disturb other FLTK development. After all, F stands for fast, L for light.


tl;dr: silly example on how had it is to implement language specific stuff.

One short example on how important it is to know the language, even if it is very similar to English, and even more so for scripts we (the current devs) know nothing about. In German, we have four additional characters to the latin alphabet: äöü and ß. Sounds easy, they are just shortcuts for ae, oe, ue and sz. We also have capitalization for ÄÖÜ, but not for ß. A capitalized ß becomes 'SS'. So one character becomes two, which is easy to implement for strings, but not for individual characters. And everything fails if we go from SS back to small letters. Is it ss now or ß? The only way to know is to know the meaning of the word, know how it used in the sentence, and know if the author writes in traditional or modernizes German. Maybe it's the name of a person? The only way to know for sure is to ask the person how they prefer their name to be written... .

**Comment by ManoloFLTK** [comment]
My reading of this issue is that Gonzalo isn't asking for an FLTK text editor widget supporting right-to-left scripts. Rather, he suggests to extend widgets such as Fl_Flex to support flowing right-to-left. That may be a realistic objective for current FLTK devs, after 1.4 gets released.

As for an FLTK text editor widget supporting right-to-left scripts, I believe it could be relatively easy to get that for macOS introducing a new widget, say Fl_Native_Text_Editor, that would use an NSTextView cocoa object laid above the single NSView FLTK creates for each Fl_Window and would inherit all multi-language, text display, edition, selection, and formatting capabilities of macOS at no cost. FLTK already uses NSView's laid over GL windows under macOS to support child widgets of Fl_Gl_Window's when OpenGL3+ is used. I wonder if that approach would be doable for Windows and Linux?

**Comment by ggarra13** [comment]
> he suggests to extend widgets such as Fl_Flex to support flowing right-to-left. That may be a realistic objective for current FLTK devs, after 1.4 gets released.

Right.  I think supporting RTL languages has to be approached in steps (milestones):

1) Modify Fl_Flex and Fl_Pack (or create new ones) that layout their widgets right to left.
2) Modify Menu layout so that submenus appear to the left.  This may be a consideration to take into account considering there were talks about rewriting the menu code.
3) Providing a RTL typing editor (as Manolo suggests).  As a Text Editor for, say, Python, it might need to switch between normal LTR code typing to RTL typing for messages.  I think how this is handled in Emacs is the example to follow.

1) seems doable to me. 2) is harder.  3) is really difficult to get right and would indeed need someone with knowledge of a RTL language.

**Comment by MatthiasWM** [comment]
1. that should not be too hard to implement
2. yes, I would like to implement menu items that derive from Fl_Button, so we would no longer need special menu code and gain all the advantages from Fl_Widget, including icon and text alignment, which would be relevant here. I am very much looking forward to implement that for 1.5 .
3. I don't use Emacs, but I am sure there is a video about that somewhere. My concern with any of the text editing widgets is, that they are quite complex, to the point where they are sometimes quite convoluted. Not sure if refactoring would fix things. But since I am relatively deep in Fl_Text_Editor, I can try things out - assuming someone can tell me what I am looking for.

@ManoloFLTK 's idea of Native text editors is a nice addition, assuming we can get the entire FLTK API projected onto the native widget.

**Comment by MatthiasWM** [comment]
Looking into this a bit more, I could add `FL_ALIGN_RTL` and `FL_MENU_RTL` as alignment flags for widgets and menu items. This would give a base for implementing layout code where FL_ALIGN_RIGHT and right image alignment is not sufficient.

Some widgets are really easy to adapt, like Fl_Light_Button, checkboxes, etc. which simply moves the indicator rectangle to the right. The same for pulldown menus. Fl_Menu_Item is also very easy to adapt. Moving scrollbars to the left is a bit more tricky, but not impossible. The hard one is Fl_Tabs because, and text input is really difficult.

So all in all, the API changes are minimal, no change to the ABI, and RTL can be implemented per widget.

The questions is: should we start that now, or wait until 1.5?

**Comment by ManoloFLTK** [comment]
We are extremely late to release 1.4. We MUST work only to have this release out. My view is that RTL can wait til 1.5.

**Comment by MatthiasWM** [comment]
> We are extremely late to release 1.4. We MUST work only to have this release out. My view is that RTL can wait til 1.5.

Yes, you are right.

**Comment by Albrecht-S** [comment]
Just some thoughts to keep in mind **_for an implementation in 1.5.0 or later_**. I'm writing this **now** because I stumbled across this issue when looking for open issues relevant for 1.4.0.

@MatthiasWM wrote:

> Looking into this a bit more, I could add FL_ALIGN_RTL and FL_MENU_RTL as alignment flags for widgets and menu items. This would give a base for implementing layout code where FL_ALIGN_RIGHT and right image alignment is not sufficient.

I don't think that this would be very helpful. I believe that _"RTL Layout"_ of widgets, labels, and menus (as we might call it) would have to be a kind of "global flag" or setting which would be established at program startup, especially if we think of `i18n` which loads the particular language setting **_depending on the locale at program startup_**. It wouldn't make sense to require the user (i.e. programmer) to set the flag in **every** menu item and widget that needs such an alignment (or "RTL Layout").

> Moving scrollbars to the left is a bit more tricky, but not impossible.

Fl_Scroll offers this already, see [docs of Fl_Scroll](https://www.fltk.org/doc-1.4/classFl__Scroll.html):
> _Use scrollbar.align(int) ( see void [Fl_Widget::align(Fl_Align)](https://www.fltk.org/doc-1.4/classFl__Widget.html#a7dcd6fc8c6108fb68a00bf922cc79022) ) : to change what side the scrollbars are drawn on._
> _If the FL_ALIGN_LEFT bit is on, the vertical scrollbar is on the left. If the FL_ALIGN_TOP bit is on, the horizontal scrollbar is on the top. ..._

@ggarra13 wrote:
> Modify Fl_Flex and Fl_Pack (or create new ones) that layout their widgets right to left.

According to what I wrote above there shouldn't be new widgets. The direction of flow would have to be determined "automagically" by the global "RTL Layout" flag (or directly by the locale or whatever).

> Modify Menu layout so that submenus appear to the left. This may be a consideration to take into account considering there were talks about rewriting the menu code.

Note that there is (currently) no strict layout requirement that submenus appear to the right. This is the default, but depending on space on the screen or window, the actual position can vary. My vision for future menu layout is similar to newer Mozilla programs (e.g. Firefox and Thunderbird) that open submenus **_in the same window_** as the parent menu. This leaves only the decision where to let the mark for submenus (an arrow or triangle) point to (left or right). I can't tell what Mozilla does with RTL languages (does the "arrow" point to the left side?).

**_Summary:_** This is all only meant as points to consider. I don't expect direct replies, this may be discussed later when we proceed with 1.5. The assumption that a global flag for "RTL Layout" is sufficient may be wrong, there may be exceptions (hopefully not!).

**Comment by ManoloFLTK** [comment]
See also discussion #1174 that proposes an implementation of a new FLTK text widget that would support text input and edition in all scripts, right-to-left included. That widget recruits a native widget on each platform.
